oci/pull.bzl

```diff
@@ -0,0 +1,385 @@
+    ],
+)
+"""
+
+_pin_sh = """\
+#!/usr/bin/env bash
+{rlocation}
+
+mediaType="$(cat $(rlocation {name}/mediaType.json))"
+echo -e "Replace your '{name}' declaration with the following:\n"
+
+cat <<EOF
+oci_pull(
+    name = "{name}",
+    digest = "sha256:{digest}",
+    image = "{image}",
+EOF
+
+[[ $mediaType == "{manifestListType}" ]] && cat <<EOF
+    # Listing of all platforms that were found in the image manifest.
+    # You may remove any that you don't use.
+    platforms = $(cat $(rlocation {name}/platforms.json)),
+EOF
+
+echo ")"
+"""
+
+def _pull_latest_impl(rctx):
+    """Download the 'latest' tag and create a repository that can produce pinning instructions"""
+    _download(rctx, "latest", "manifest_list.json", "manifests")
+    result = rctx.execute(["shasum", "-a", "256", "manifest_list.json"])
+    if result.return_code:
+        msg = "shasum failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr)
+        fail(msg)
+    rctx.file("pin.sh", _pin_sh.format(
+        name = rctx.attr.name,
+        digest = result.stdout.split(" ", 1)[0],
+        image = rctx.attr.image,
+        rlocation = BASH_RLOCATION_FUNCTION,
+        manifestListType = _MANIFEST_LIST_TYPE,
+    ), executable = True)
+    rctx.file("BUILD.bazel", _latest_build)
+
+pull_latest = repository_rule(
+    _pull_latest_impl,
+    attrs = {
+        "image": attr.string(doc = "The name of the image we are fetching, e.g. gcr.io/distroless/static", mandatory = True),
+    },
+)
+
+# Note: there is no exhaustive list, image authors can use whatever name they like.
+# This is only used for the oci_alias rule that makes a select() - if a mapping is missing,
+# users can just write their own select() for it.
+_DOCKER_ARCH_TO_BAZEL_CPU = {
+    "amd64": "@platforms//cpu:x86_64",
+    "arm": "@platforms//cpu:arm",
+    "arm64": "@platforms//cpu:arm64",
+    "ppc64le": "@platforms//cpu:ppc",
+    "s390x": "@platforms//cpu:s390x",
+}
+
+def oci_pull(name, image, platforms = None, digest = None):
+    """Repository macro to fetch image manifest data from a remote docker registry.
+
+    Args:
+        name: repository with this name is created
+        image: the remote image without a tag, such as gcr.io/bazel-public/bazel
+        platforms: for multi-architecture images, a dictionary of the platforms it supports
+            This creates a separate external repository for each platform, avoiding fetching layers.
+        digest: the digest string, starting with "sha256:", "sha512:", etc.
+            If omitted, instructions for pinning are provided.
+    """
+
+    if digest == None:
+        pull_latest(name = name, image = image)
+
+        # Print a command - in the future we should print a buildozer command or
+        # buildifier: disable=print
+        print("""
+WARNING: for reproducible builds, a digest is recommended.
+Run the following command to fix this:
+
+bazel run @{}//:pin
+""".format(name))
+        return
```

Comment:
I think we could do better here. I don't know if this is done to prevent users from making mistakes but I feel like this would have a smoother dx if we just resolved the digest corresponding to the tag (not just `latest`) and continue executing.

Reply:
done - I added distroless/python3 with a "debug" tag in the WORKSPACE to demonstrate, how does that look?